### ISSUE TITLE
Proposal: Remove [] from extended trait change syntax in future major release

### ISSUE THREAD

**Issue by corranwebster** [enthought/traits#539]
Extended trait name handlers which use the `[]` syntax (as in `foo.bar.baz[]`) are anomalous:

- they can only be used at the end of an expression
- they overload `old` and `new` to return `removed` and `added` but with no information about where the changes take place, which reduces utility compared to `TraitListEvent`s.
- they require lots of special-casing code
- for simply detecting that a list trait has changed, a listener with no arguments to the un-augmented extended trait (eg. `foo.bar.baz` will fire under the same circumstance), _except_ for simple traits (and in that case you can just listen for `foo` and `foo_items` together)

In short, they don't add much, and add a lot of complexity to the codebase.

This change might be coupled with making listeners which have no arguments which are listening to `List` (and similar) traits fire on `_items` changes in the same way as extended traits.

The main argument against this change is the body of existing code which uses `[]` extended traits.  A quick grep shows only a handful of uses in the ETS codebase outside of tests.

**Comment by mdickinson** [comment]
I've seen `[]` uses in project code, so this would definitely cause breakage. *If* we decide to do this, I don't think we can deprecate now and remove in Traits 6.0.0; that's too short a warning period. We could potentially deprecate in Traits 6.0.0.

**Comment by mdickinson** [comment]
I'd also suggest that this be done as part of a larger review and rework of the change notifier syntax; it would be painful to be making repeated breaking changes, as opposed to a one-time transition.

**Comment by kitchoi** [comment]
The mini-language used by `observe` won't have this `[]`, and the corresponding use case is supported by the keyword `"items"`. `[]` is still around for `on_trait_change`.
Is this issue resolved then?

**Comment by mdickinson** [comment]
Yes, let's close.
